chemprop/cli/hyperopt.py

```diff
@@ -1,16 +1,348 @@
+import json
+import logging
+import sys
 from argparse import ArgumentParser, Namespace
+from pathlib import Path
 
+import torch
+from lightning import pytorch as pl
+from lightning.pytorch.loggers import CSVLogger, TensorBoardLogger
+from lightning.pytorch.callbacks import ModelCheckpoint, EarlyStopping
+from ray import tune
+from ray.train import CheckpointConfig, RunConfig, ScalingConfig
+from ray.train.lightning import (RayDDPStrategy, RayLightningEnvironment,
+                                 RayTrainReportCallback, prepare_trainer)
+from ray.train.torch import TorchTrainer
+from ray.tune.schedulers import ASHAScheduler
+
+from chemprop.cli.common import (add_common_args, process_common_args,
+                                 validate_common_args)
+from chemprop.cli.train import (add_train_args, build_datasets, build_model,
+                                build_splits, normalize_inputs,
+                                process_train_args, validate_train_args)
 from chemprop.cli.utils.command import Subcommand
+from chemprop.data import MolGraphDataLoader
+from chemprop.nn import AggregationRegistry
+from chemprop.nn.utils import Activation
+
+
+logger = logging.getLogger(__name__)
+
+AVAILABLE_SPACES = {
+    "activation": tune.choice(list(Activation.keys())),
+    "aggregation": tune.choice(list(AggregationRegistry.keys())),
+    "aggregation_norm": tune.quniform(lower=1, upper=200, q=1),
+    "batch_size": tune.quniform(lower=5, upper=200, q=5),
+    "depth": tune.randint(lower=2, upper=6),
+    "dropout": tune.choice(
+        [
+            tune.choice([0.]),
+            tune.quniform(lower=0.05, upper=0.4, q=0.05),
+        ],
```

Comment:
Why is the dropout ratio defined this way? I notice the code here seems to sample a dropout ratio of 0 more often than other choices. I understand that Hyperopt package may encounter issues in finding a suitable dropout ratio in v1, sometimes consistently choosing a large dropout rate, even though a lower dropout ratio is usually better in our experience. However, could this also be a problem with Ray Tune so we need to design it this way?

Reply:
I don't think this samples `0` more often? This should basically just resolve to `tune.choice([0, 0.05, 0.10, 0.15, 0.20, 0.25, 0.3, 0.35])` (which, actually, why don't we just do that instead?)

Reply:
This code defines a dropout rate that can either be exactly 0 or a value between 0.05 and 0.4, chosen in steps of 0.05. It uses `tune.choice` for selecting between these two options, and `tune.quniform` for specifying the range and step size for the second option.
I think `tune.choice([0, 0.05, 0.10, 0.15, 0.20, 0.25, 0.3, 0.35])` is effectively the same as `tune.quniform(lower=0, upper=0.4, q=0.05)`, which is what we did in v1.

Reply:
@shihchengli @JacksonBurns This follows Chas's PR for v1 here: https://github.com/chemprop/chemprop/pull/571/files. He did a thorough study on a better default search range.